### PR TITLE
perf: 120x faster clone with real-world performance tests

### PR DESCRIPTION
**Pull request by OmarIthawi**
## Description
`git add remote --fetch origin` is very slow for any medium to large repository. With this fix, the performance is hugely improved by fetching only after sparse-checkout rules are set, therefore making checkout almost instant.

Additionally, the `git` steps are simplified into fewer steps.

## Testing

 - [x] Added an integration (real) test scenario with timing constraint of 10 seconds for the `edx-platform:conf/locale` directory
 - [x] Fix existing broken test cases after initial review

## Test results
The `new_` implementation is waaaay faster!

| Test | Old | New  |
|--------|--------|--------|
| Clone edx-ios-app Source directory | 1m 57s | 1s |
| Clone edx-platform | Too much | 2s |

<details><summary>Performance testing code</summary>
<p>

```shell
new_clone_ios() {
  cd /tmp
  rm -rf /tmp/atlas-repo
  mkdir -p /tmp/atlas-repo
  cd /tmp/atlas-repo
  git clone --filter=blob:none --no-checkout --depth=1 --branch=master "https://github.com/openedx/edx-app-ios.git" atlas-repo
  cd atlas-repo
  git sparse-checkout set '!*'
  git sparse-checkout add 'Source/**'
  git checkout HEAD
  tree
}

old_clone_ios() {
  cd /tmp
  rm -rf /tmp/atlas-repo
  mkdir -p /tmp/atlas-repo
  cd /tmp/atlas-repo
  git init -b master
  git remote add -f origin "https://github.com/openedx/edx-app-ios.git"
  git sparse-checkout set '!*'
  git sparse-checkout add 'Source/**'
  git pull origin master
  tree
}

new_clone_edx_platform() {
  cd /tmp
  rm -rf /tmp/atlas-repo
  mkdir -p /tmp/atlas-repo
  cd /tmp/atlas-repo
  git clone --filter=blob:none --no-checkout --depth=1 --branch=master "https://github.com/openedx/edx-platform.git" atlas-repo
  cd atlas-repo
  git sparse-checkout set '!*'
  git sparse-checkout add 'conf/locale/ar**'
  git checkout HEAD
  tree
}

old_clone_edx_platform() {
  cd /tmp
  rm -rf /tmp/atlas-repo
  mkdir -p /tmp/atlas-repo
  cd /tmp/atlas-repo
  git init -b master
  git remote add -f origin "https://github.com/openedx/edx-platform.git"
  git sparse-checkout set '!*'
  git sparse-checkout add 'conf/locale/ar**'
  git pull origin master
  tree
}


##################
time new_clone_ios
# real	0m1.147s
# user	0m0.064s
# sys	0m0.016s

##################
time old_clone_ios
# real	1m57.448s
# user	0m35.059s
# sys	0m9.742s

###################
time new_clone_edx_platform
# real	0m2.600s
# user	0m0.237s
# sys	0m0.079s

##################
time old_clone_edx_platform
# real infinity!
# user infinity!
# sys infinity!

```

</p>
</details> 

References
----------

This pull request is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which is sparked by the [Translation Infrastructure update OEP-58](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0058-arch-translations-management.html#specification).

Check the links above for full information about the overall project.

Internalization is being rearchitected in Open edX Python, XBlock, Micro-frontend, and other projects. There are a number of immediately visible changes:
 - Remove source and language translations from the repositories, hence no `.json`, `.po` or `.mo` files will be committed into the repos.
 - Add standardized `make extract_translations` in all repositories
 - Push user-facing messages strings into [openedx/openedx-translations](https://github.com/openedx/openedx-translations/).
 - Integrate root repositories with [openedx/openedx-atlas](https://github.com/openedx/openedx-atlas/) to pull translations on build/deploy time

Breaking Changes
----------------

One of the primary goals of the project is to **avoid breaking changes**. If you noticed any suspicious code, please raise your concern. But before that, please know the strategy we're following to avoid breaking changes.

### CODE REVIEW

**Review by brian-smith-tcril — COMMENTED** [review]
This looks like a very reasonable approach to me, and the 120x performance boost is incredible!

Once the tests are passing I'll give it another look, but overall I'm very happy with it!

**Review by brian-smith-tcril — COMMENTED** [review]
Overall this looks great! One tiny suggestion and then this should be good to merge!

File: spec/verbosity_spec.sh [hunk 1]
```diff
@@ -23,6 +23,14 @@ Describe 'Test default verbosity'
       %preserve GIT_CALLED_NOT_VERBOSELY
     }
 
+    cd() {
+      true # Omit output
+    }
+
+    rm() {
+      true # Omit output
+    }
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply spec/verbosity_spec.sh]: Done. I somehow missed that pattern. Thanks.

File: atlas [hunk 2]
```diff
@@ -336,7 +334,7 @@ pull_translations() {
   # Reset git sparse-checkout list of include/exclude files
   # Tells sparse checkout to ignore all files, except when otherwise noted
   # See https://www.git-scm.com/docs/git-sparse-checkout for detailed implementation
-  git sparse-checkout set "!*"
+  git sparse-checkout set --no-cone "!*"
```
Comment (OmarIthawi) [inline atlas:337]: `--no-cone` ensures compatilbity with git 2.40+

**Review by brian-smith-tcril — APPROVED** [review]
Beautiful! LGTM :shipit:

**Comment by openedx-webhooks** [comment]
@OmarIthawi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-atlas&entry.867055334=https://github.com/openedx/openedx-atlas/pull/14&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2023-04-07+12:54&entry.1917517419=2023-04-11+11:46&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**